File: section1/unit3/exercise19.clj
(ns exercise19
  (load-file "list-node.clj"))

(import unit3.list_node.ListNode)

(def c (ListNode. 3 nil))
(def b (ListNode. 2 c))
(def a (ListNode. 1 b))

(println (:val a))
(println (-> a :next :next :val))

(def level [])
(def head a)
(while (boolean (:next a)) (do
  (def level (conj level :next))
  (def a (:next a))))

(try
  (println level)
  (def head (assoc-in head level nil))
  (println head)
(catch Exception e
  (println "caught exception" (.getMessage e))))

File: section1/unit3/exercise19.go
// +build exercise19

package main

import "fmt"

type listNode struct {
	Val  int
	Next *listNode
}

func deleteTail(h *listNode) {
	tmp := h
	for ; h.Next != nil; h = h.Next {
		tmp = h
	}
	tmp.Next = nil
}

func main() {
	f := listNode{6, nil}
	e := listNode{5, &f}
	d := listNode{4, &e}
	c := listNode{3, &d}
	b := listNode{2, &c}
	a := listNode{1, &b}
	fmt.Println(a.Val)
	fmt.Println(a.Next)
	fmt.Println(a.Next.Next.Next.Next.Next)
	deleteTail(&a)
	fmt.Println(a)
	fmt.Println(a.Next)
	fmt.Println(a.Next.Next.Next.Next.Next)
}
